### PR TITLE
Widget: Add tests for backcompat

### PR DESCRIPTION
**Pull request by gabrielschulhof**
Fixes gh-7972

Also modifies backcompat somewhat:
- We cannot treat the values of class keys as truthy, because `""` counts as true, because a key may be explicitly set to `""` - for example, if `"ui-widgetname": "ui-corner-all"` and you don't want corners
- `Array.splice()` is incapable of removing the first element of an array, so we must use `Array.shift()` in that special case
- When we compute the initial value of the options during _create(), we cannot interleave the precedence of the value of `this.options.classes[ this.classProp ]` with the precedence of the deprecated style option values. We must assume that the initial value of the class key reflects the default values of the style options. We also cannot modify both the value of the class key and the value of the options, because that results in unexpected behaviour.
  We must instead assume that, if `this.options.classes[ this.classProp ]` has been modified, then the application we're a part of is running non-deprecated code and the value of `this.options.classes[ this.classProp ]` completely determines the initial value of the style options.
  Conversely, if `this.options.classes[ this.classProp ]` is not modified (i.e., equal to the value in the prototype), then we must assume that legacy code is in place, and we must use the value of the various style options to construct `this.options.classes[ this.classProp ]`

### CODE REVIEW

**Comment by arschmitz** [comment]
@gabrielschulhof 
> We must instead assume that, if this.options.classes[ this.classProp ] has been modified, then the application we're a part of is running non-deprecated code and the value of this.options.classes[ this.classProp ] completely determines the initial value of the style options. Conversely, if this.options.classes[ this.classProp ] is not modified (i.e., equal to the value in the prototype), then we must assume that legacy code is in place, and we must use the value of the various style options to construct this.options.classes[ this.classProp ]

We cannot make this assumption because deprecated and non deprecated code can be used at the same time for example someone may be using a plugin that uses old code while they themselves updated to new API in their app.

**Comment by gabrielschulhof** [comment]
A more detailed discussion: http://irc.jquery.org/%23jqueryui-dev/default_%23jqueryui-dev_20150223.log.html#t07:12:23

File: js/widgets/widget.backcompat.js [hunk 1]
```diff
@@ -23,12 +39,12 @@ if ( $.mobileBackcompat !== false ) {
...
 
 		_create: function() {
-			this._setInitalOptions();
+			this._setInitialOptions();
 			this._super();
 		},
 
 		_enhance: function() {
-			if ( !this.options.enhanced && this.options.wrapperClass ) {
+			if ( this.options.wrapperClass ) {
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: Why was this changed?

Comment (gabrielschulhof) [inline js/widgets/widget.backcompat.js]: Because _enhance() is only called if the widget is not pre-rendered.

Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: Thats not true.

File: js/widgets/widget.backcompat.js [hunk 2]
```diff
@@ -51,50 +63,62 @@ if ( $.mobileBackcompat !== false ) {
...
+		_optionsToClasses: function( option, value ) {
 			var classArray,
-				newValue = "",
 				prop = this.classProp,
 				classes = this.options.classes,
 				className = this._boolOptions[ option ];
 
 			if ( prop ) {
-				classArray = classes[ prop ].split ( " " );
+				classArray = classes[ prop ] ? classes[ prop ].split ( " " ) : [];
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: Use `.match( /\S+/g )` instead of `.split( " " ) `

File: js/widgets/widget.backcompat.js [hunk 3]
```diff
@@ -51,50 +63,62 @@ if ( $.mobileBackcompat !== false ) {
...
+						}
+					}) ;
 
-						if ( initial !== current ) {
-							options[ option ] = current;
-						} else if ( options[ option ] !== original[ option ] ) {
-							that._optionsToClasses( option, options[ option ] );
+				// Otherwise we assume that we're dealing with legacy code and look for style
+				// option values which diverge from the defaults, and update the classes option
+				// value accordingly
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: This is not true there is the third case where they are both default where we should do nothing.

File: js/widgets/widget.backcompat.js [hunk 4]
```diff
@@ -51,50 +63,62 @@ if ( $.mobileBackcompat !== false ) {
...
+				} else {
+					$.each( this._boolOptions, function( option, className ) {
+						if ( options[ option ] !== original[ option ] ) {
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: Your covering the third case i mention above here, however thats not what your comment says.

File: tests/unit/widget/backcompat_core.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+		shadow: true,
+		corners: false
+	},
+	classProp: "test-testwidget",
+	_create: function() {
+		this._addClass( this.element, "test-testwidget" );
+		this._enhance();
+	},
+	_enhance: $.noop
+} );
```
Comment (arschmitz) [inline tests/unit/widget/backcompat_core.js]: Your polluting the global test scope here and adding a new namespace on `$` this should be in the module setup and use the mobile NS. In teardown
```js
if ( $.mobile ) {
	delete $.mobile.testwidget;
	delete $.fn.testwidget;
}
```

**Comment by arschmitz** [comment]
Much of what your doing in your backcompat tests replicates whats already in, https://github.com/jquery/jquery-ui/blob/master/tests/unit/testsuite.js, rather then recreate the wheel can you look at just pulling in that test file. then we can add tests like https://github.com/jquery/jquery-ui/blob/classes/tests/unit/dialog/dialog_common.js and https://github.com/jquery/jquery-ui/blob/classes/tests/unit/dialog/dialog_common_deprecated.js

**Comment by gabrielschulhof** [comment]
@arschmitz AFAICT testsuite.js tests plugins rather than instances. In fact, it doesn't instantiate any widgets. OTOH I'm checking for specific behaviour within a single instance. Maybe it's just my lack of JS-fu that I'm not seeing the overlap :/

File: tests/unit/widget/backcompat_core.js [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+				this._addClass( this.element, "test-testwidget" );
+				this._enhance();
+			},
+			_enhance: $.noop
+		} );
+		$.widget( "mobile.testwidget", $.mobile.testwidget, $.mobile.widget.backcompat );
+	},
+	teardown: function() {
+		delete $.mobile.testwidget;
+		if ( !this.hadMobile ) {
```
Comment (arschmitz) [inline tests/unit/widget/backcompat_core.js]: This cant ever be true? Also why are you doing this?

File: tests/unit/widget/backcompat_core.js [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+			corners: true
+		},
+		"Widget options reflect classes option values" );
+
+	assert.hasClasses( testWidget[ 0 ], "ui-corner-all", "Element has corner class" );
+	assert.lacksClasses( testWidget[ 0 ], "ui-button-inline ui-mini ui-shadow",
+		"Element does not have inline, mini, and shadow classes" );
+
+	testWidget.testwidget( "option", "classes.test-testwidget", "ui-button-inline ui-mini" );
+
```
Comment (arschmitz) [inline tests/unit/widget/backcompat_core.js]: remove blank line and group the manipulation with the test for it in a block

**Comment by arschmitz** [comment]
In the tests file can you remove the extra blank lines and group the tests together in logical groups the the accompanying manipulations for that test.

File: tests/unit/widget/backcompat_core.js (hunk 6 above)
Comment (gabrielschulhof) [inline tests/unit/widget/backcompat_core.js]: I just realized that there's no reason for it, since I'm expecting $.mobile.widget.backcompat to be present. *facepalm*

File: js/widgets/widget.backcompat.js [hunk 8]
```diff
@@ -23,22 +41,18 @@ if ( $.mobileBackcompat !== false ) {
...
 
 		_create: function() {
-			this._setInitalOptions();
+			this._setInitialOptions();
 			this._super();
-		},
-
-		_enhance: function() {
 			if ( !this.options.enhanced && this.options.wrapperClass ) {
 				this.widget().addClass( this.options.wrapperClass );
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: this should use `this._addClass( this.widget(), this.options.wrapperClass );`

File: js/widgets/widget.backcompat.js [hunk 9]
```diff
@@ -116,6 +142,12 @@ if ( $.mobileBackcompat !== false ) {
 			}
 
 			this._superApply( arguments );
+		},
+
+		destroy: function() {
+			if ( this.options.wrapperClass && !this.options.enhanced ) {
+				this.widget().removeClass( this.options.wrapperClass );
+			}
 		}
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: This whole method can be dropped once you fix the addition of the class above. Also you should never use the "destroy" method this is part of the base widget always use `_destroy` which is called from `destroy` internally in the base widget. The above code will break the `destroy` method in every widget that uses this.

File: js/widgets/widget.backcompat.js [hunk 10]
```diff
@@ -13,6 +13,24 @@ define( [
 (function( $, undefined ) {
 
 if ( $.mobileBackcompat !== false ) {
+
+	var classSplitterRegex = /\S+/g,
+
+		// splice() is incapable of removing the first element, because specifying a negative
+		// starting index is interpreted as wanting to count from the end of the array. So, we use
+		// shift() to remove the first element
+		removeOneElementFromArray = function( array, index ) {
```
Comment (arschmitz) [inline js/widgets/widget.backcompat.js]: Lets use a slightly less verbose name here. Maybe just `removeFromArray` ?

File: tests/unit/widget/backcompat_core.js [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+} );
+
+test( "Style options are updated when the classes option changes", function( assert ) {
+	var testWidget = $( "<div>" );
+
+	testWidget.testwidget( {
+		classes: {
+			"test-testwidget": "ui-corner-all"
+		}
+	} );
```
Comment (arschmitz) [inline tests/unit/widget/backcompat_core.js]: Just do `var testWidget = $.mobile.testWidget();` then you can do `testWidget.option( "classes", ...` and `testWidget.option() `, `testWidget.element[ 0 ]` etc below

File: tests/unit/widget/backcompat_core.js [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+			shadow: false,
+			corners: false
+		},
+		"Widget options reflect classes option values" );
+	assert.hasClasses( testWidget[ 0 ], "ui-button-inline ui-mini" );
+	assert.lacksClasses( testWidget[ 0 ], "ui-shadow ui-corner-all" );
+} );
+
+test( "The classes option is updated when style options change", function( assert ) {
+	var testWidget = $( "<div>" );
```
Comment (arschmitz) [inline tests/unit/widget/backcompat_core.js]: Same here as in previous test

**Comment by arschmitz** [comment]
@gabrielschulhof can you update the tests to instantiate like I commented on the first then let me know and ill finish reviewing the tests.

File: tests/unit/widget/backcompat_core.js (hunk 11 above)
Comment (gabrielschulhof) [inline tests/unit/widget/backcompat_core.js]: Nice!

**Comment by gabrielschulhof** [comment]
I'll drop the test that checks whether wrapperClass behaves correctly in conjunction with pre-rendered widgets. The widget backcompat should not be enhanced-option-aware, because we're doing that per-widget.

**Comment by arschmitz** [comment]
ok looks good now i think :+1:
